sky/data/sky_csync.py

```diff
@@ -0,0 +1,337 @@
+    _CONN.commit()
+
+
+@connect_db
+def _get_csync_pid_from_source_path(path: str) -> Optional[int]:
+    """Given the path, returns process ID of csync running on it"""
+    assert _CURSOR is not None
+    _CURSOR.execute(
+        'SELECT csync_pid FROM running_csync '
+        'WHERE source_path=(?) AND boot_time=(?)', (path, _BOOT_TIME))
+    row = _CURSOR.fetchone()
+    if row:
+        return row[0]
+    return None
+
+
+@click.group()
+def main():
+    pass
+
+
+def get_s3_upload_cmd(src_path: str, dst: str, num_threads: int, delete: bool,
+                      no_follow_symlinks: bool):
+    """Builds sync command for aws s3"""
+    config_cmd = ('aws configure set default.s3.max_concurrent_requests '
+                  f'{num_threads}')
+    subprocess.check_output(config_cmd, shell=True)
+    sync_cmd = f'aws s3 sync {src_path} s3://{dst}'
+    if delete:
+        sync_cmd += ' --delete'
+    if no_follow_symlinks:
+        sync_cmd += ' --no-follow-symlinks'
+    return sync_cmd
+
+
+def get_gcs_upload_cmd(src_path: str, dst: str, num_threads: int, delete: bool,
+                       no_follow_symlinks: bool):
+    """Builds sync command for gcp gcs"""
+    sync_cmd = (f'gsutil -m -o \'GSUtil:parallel_thread_count={num_threads}\' '
+                'rsync -r')
+    if delete:
+        sync_cmd += ' -d'
+    if no_follow_symlinks:
+        sync_cmd += ' -e'
+    sync_cmd += f' {src_path} gs://{dst}'
+    return sync_cmd
+
+
+def run_sync(src: str, storetype: str, dst: str, num_threads: int,
+             interval_seconds: int, delete: bool, no_follow_symlinks: bool,
+             csync_pid: int):
+    """Runs the sync command to from src to storetype bucket"""
+    #TODO(Doyoung): add enum type class to handle storetypes
+    storetype = storetype.lower()
+    if storetype == 's3':
+        sync_cmd = get_s3_upload_cmd(src, dst, num_threads, delete,
+                                     no_follow_symlinks)
+    elif storetype == 'gcs':
+        sync_cmd = get_gcs_upload_cmd(src, dst, num_threads, delete,
+                                      no_follow_symlinks)
+    else:
+        raise ValueError(f'Unsupported store type: {storetype}')
+
+    max_retries = 10
+    # interval_seconds/2 is heuristically determined
+    # as initial backoff
+    initial_backoff = int(interval_seconds / 2)
+    backoff = common_utils.Backoff(initial_backoff)
+    for _ in range(max_retries):
+        try:
+            with subprocess.Popen(sync_cmd, start_new_session=True,
+                                  shell=True) as proc:
+                _set_running_csync_sync_pid(csync_pid, proc.pid)
+                if storetype == 's3':
+                    # set number of threads back to its default value
+                    config_cmd = \
+                        ('aws configure '
+                         'set default.s3.max_concurrent_requests 10')
+                    subprocess.run(config_cmd, shell=True, check=True)
```

Comment:
Why are we setting those to num_threads first, and then set back here?

Reply:
This part was originally at the end of `run_sync`, but moved it closer to `Popen` trying to lower the chance of user separately runs aws s3 cli command at the same time. to reflect on your concern:

> What will happen if two such commands run simultaneously? IIUC the second will cover the modification for the first command?

Since `Popen` already ran the command by the time it reaches this portion, sync command runs with `num_thread` user provided first, and then we set back to default setting.

Reply:
Just updated so that all three commands to thread set to user's requirement, running sync command, thread set back to default are wrapped into a single command separated by `';'` at `get_s3_upload_cmd` and ran with `subprocess.Popen()` altogether at `run_sync`

Reply:
Also, updated the `get_s3/gcs_upload_cmd` using list to build commands rather than string concatenations for cleaner implementation.